src/EditorFeatures/Core/Implementation/ReferenceHighlighting/ReferenceHighlightingViewTaggerProvider.cs

```diff
@@ -2,87 +2,196 @@
 {
     [Export(typeof(IViewTaggerProvider))]
     [ContentType(ContentTypeNames.RoslynContentType)]
-    [TagType(typeof(AbstractNavigatableReferenceHighlightingTag))]
+    [TagType(typeof(NavigableHighlightTag))]
     [TextViewRole(PredefinedTextViewRoles.Interactive)]
-    internal partial class ReferenceHighlightingViewTaggerProvider :
-        AbstractAsynchronousViewTaggerProvider<AbstractNavigatableReferenceHighlightingTag>
+    internal partial class ReferenceHighlightingViewTaggerProvider : AsynchronousTaggerProvider<NavigableHighlightTag>
     {
         private readonly ISemanticChangeNotificationService _semanticChangeNotificationService;
 
+        // Whenever an edit happens, clear all highlights.  When moving the caret, preserve 
+        // highlights if the caret stays within an existing tag.
+        public override TaggerCaretChangeBehavior CaretChangeBehavior => TaggerCaretChangeBehavior.RemoveAllTagsOnCaretMoveOutsideOfTag;
+        public override TaggerTextChangeBehavior TextChangeBehavior => TaggerTextChangeBehavior.RemoveAllTags;
+        public override IEnumerable<PerLanguageOption<bool>> PerLanguageOptions => SpecializedCollections.SingletonEnumerable(FeatureOnOffOptions.ReferenceHighlighting);
+
         [ImportingConstructor]
         public ReferenceHighlightingViewTaggerProvider(
             IForegroundNotificationService notificationService,
             ISemanticChangeNotificationService semanticChangeNotificationService,
-            [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners) :
-            base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.ReferenceHighlighting), notificationService)
+            [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners)
+            : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.ReferenceHighlighting), notificationService)
         {
             _semanticChangeNotificationService = semanticChangeNotificationService;
         }
 
-        protected override bool RemoveTagsThatIntersectEdits => true;
+        public override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
+        {
+            // Note: we don't listen for OnTextChanged.  Text changes to this this buffer will get
+            // reported by OnSemanticChanged.
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnCaretPositionChanged(textView, textView.TextBuffer, TaggerDelay.Short),
+                TaggerEventSources.OnSemanticChanged(subjectBuffer, TaggerDelay.OnIdle, _semanticChangeNotificationService),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.Short),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, FeatureOnOffOptions.ReferenceHighlighting, TaggerDelay.NearImmediate));
```

Comment:
If you specify the option in PerLanguageOptions, should we just automatically add the event source rather than duplicating this?